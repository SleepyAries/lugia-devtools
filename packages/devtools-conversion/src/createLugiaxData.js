/**
 *
 * create by ligx
 *
 * @flow
 */
export const LugiaxDataPrefix = '$__data__';

export function getImportScripts(scripts: Object): string {
  if (!scripts) {
    return '';
  }
  const importCodes = [];

  Object.keys(scripts).forEach((codeName: string): void => {
    const { type = 'Code', code = '' } = scripts[codeName];
    if (!code) {
      return;
    }
    switch (type) {
      case 'Project':
        return importCodes.push(`import __data__${codeName} from ${code}`);
      default:
    }
  });
  return importCodes.join(';');
}
function getScripts(scripts: Object): string {
  if (!scripts) {
    return '';
  }
  const codes = [];
  Object.keys(scripts).forEach((codeName: string) => {
    const { type = 'Code', code = '' } = scripts[codeName];
    if (!code) {
      return;
    }
    switch (type) {
      case 'Code':
        codes.push(
          `const ${LugiaxDataPrefix}${codeName} = new Function('param', \`${code}\`);`
        );
        return;
      default:
    }
  });
  return codes.join('');
}
export default function(lugiax: Object): string {
  const { pageData } = lugiax;

  if (!pageData) {
    return '';
  }
  const { model } = pageData;
  if (!model) {
    return '';
  }
  const { state = {}, scripts } = pageData;

  return `const ${LugiaxDataPrefix} = LugiaxData.createData({
    model: '${model}',
    state: ${JSON.stringify(state)},
  });
  ${getScripts(scripts)}
 `;
}
