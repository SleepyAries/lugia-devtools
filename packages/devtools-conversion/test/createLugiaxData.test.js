/**
 *
 * create by guorg
 *
 * @flow
 */
import createLugiaxData, { getImportScripts } from '../src/createLugiaxData';

describe('conversion createLugiaxData. ->', () => {
  it('createLugiaxData pageData is empty', () => {
    expect(createLugiaxData({})).toBe('');
  });
  it('createLugiaxData pageData model is empty str', () => {
    expect(
      createLugiaxData({
        model: '',
      })
    ).toBe('');
  });
  it('createLugiaxData pageData model hello state', () => {
    expect(
      createLugiaxData({
        pageData: {
          model: 'hello',
        },
      })
    ).toMatchSnapshot();
  });
  it('createLugiaxData pageData model hello state {num, str, bool, date}', () => {
    expect(
      createLugiaxData({
        pageData: {
          model: 'hello',
          state: { num: 1, str: 'abc', bool: true },
        },
      })
    ).toMatchSnapshot();
  });
  it('createLugiaxData pageData model hello state {num, str, bool, date} scriptes', () => {
    expect(
      createLugiaxData({
        pageData: {
          model: 'hello',
          state: { num: 1, str: 'abc', bool: true },
          scripts: {
            Code1: {
              code: '',
            },
            Code2: {
              code: 'console.info(param);',
            },
            Code3: {
              type: 'Project',
              code: '../lugiax',
            },
          },
        },
      })
    ).toMatchSnapshot();
  });
  it('getImportScripts', () => {
    expect(
      getImportScripts({
        Code1: {
          code: '',
        },
        Code2: {
          code: 'console.info(param);',
        },
        Code3: {
          type: 'Project',
          code: '../lugiax',
        },
      })
    ).toMatchSnapshot();
  });
});
