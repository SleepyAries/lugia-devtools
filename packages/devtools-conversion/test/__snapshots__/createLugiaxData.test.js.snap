// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conversion createLugiaxData. -> createLugiaxData pageData model hello state {num, str, bool, date} 1`] = `""`;

exports[`conversion createLugiaxData. -> createLugiaxData pageData model hello state {num, str, bool, date} scriptes 1`] = `
"const __data__ = LugiaxData.createModel({
    model: 'hello',
    state: {\\"num\\":1,\\"str\\":\\"abc\\",\\"bool\\":true},
  });
  const __data__Code2 = new Function('param', 'console.info(param);');
 "
`;

exports[`conversion createLugiaxData. -> createLugiaxData pageData model hello state 1`] = `
"const __data__ = LugiaxData.createModel({
    model: 'hello',
    state: {},
  });
  
 "
`;

exports[`conversion createLugiaxData. -> getImportScripts 1`] = `"import Code3 from ../lugiax"`;
